Fixes things about goliaths: wallhacks/range hacks(no, really) and tentacles not spawning in mineral turfs; also fixes `find_potential_targets` wallhacks

Sealed101 opened the pull request:
## About The Pull Request

Goliath's sand digging behaviour could potentially target a turf that's actually unreachable by the goliath, e.g.
```
G#
#T
```
where G - goliath # - wall T - target turf. fixed that, but i think there could be something easier here, maybe instead grabbing turfs in goliath's `view()`? unsure

The component goliaths use to telegraph their attacks (`basic_mob_attack_telegraph`) casts a `do_after()` to perform the attack, but it was not actually checking for the target staying in melee range, as it was using the source goliath as both `user` and `target`, so it didn't actually care at all for the target. Implemented an `extra_checks` to `Adjacent()` since that's the closest we get for melee range shenanigans I suppose
This still allows the source basicmob to attack the target if the target moves around the source basicmob.

`!`Goliaths were also able to summon tentacles on a target that moved into cover and still stayed in the `find_potential_targets` target range. Which meant more wallhacks. This was a thing for the base `find_potential_targets`, meaning that every basic mob using it was a dirty haxxor (or very vengeful). Fixed that by making `find_potential_targets` also check for `can_see()` before proceeding further down `find_potential_targets/perform()`. `!` The only exception to this check currently are bileworms.

`!`Goliath tentacles were not spawning in mineral turfs as their `Initialize()` checked for closed turfs before handling mineral turf mining. Fixed that as well.

## Why It's Good For The Game
![Dr__Hax_by_Didgeridoo_Dealer](https://github.com/tgstation/tgstation/assets/75863639/fbcbfc1b-f489-435e-bb01-677f55398787)

## Changelog

:cl:
fix: fixed goliaths digging sand that they can't actually reach (behind windows or inbetween closed turfs)
fix: fixed goliaths melee attacking their target despite the target running away from goliath melee range
fix: fixed goliath tentacles not spawning in mineral turfs
fix: fixed goliaths summoning tentacles on targets that moved behind cover but stayed in their targeting range. this applies for most basic mobs, really, so if any basic mob was targeting you despite you hauling ass behind cover, they shouldn't anymore
/:cl:

Sealed101 committed:
fixes two things about goliaths
fix goliaths digging unreachable turfs
e.g. G - goliath # - wall T - target turf
G#
#T
or directional windows for example

fixes basic_mob_attack_telegraph not clearing up the attack do_after when target leaves source mob's adjacent turfs
it was using source as both user and target for the do_after
and it wasn't actually checking for adjacency
this still allows the source basicmob to punch the target if they move to a turf that's adjacent to the source basicmob

Sealed101 committed:
makes targeting care for seeing the target
which makes targetting datums properly clean up target_key if target is out of our range
could this be better placed in can_attack instead? not sure

Sealed101 commented:
goliaths were also reported (and witnessed (and killed me) by me) to be able to tentacle their target if it was in the targeting range, but not in sight. this was actually the base behavior of the `find_potential_targets` datum, so the fix in the latest commit will affect almost every basic mob, since i'd imagine most of em use `find_potential_targets` in some way or another.

for goliaths this means that they won't wallhack with tentacles and walk at their target despite the target being behind a wall; they will forget the target immediately as it leaves their line of sight.

i feel like that isn't the best solution and that it kind of dumbs down the mob's targeting capabilities but i also feel that wallhacks are bullshit


https://github.com/tgstation/tgstation/assets/75863639/c1bb5824-f527-4917-897e-5ccba724ff69

Sealed101 committed:
fixes tentacles not spawning in mineral turfs

Sealed101 commented:
tentacles spawn on mineral turfs properly now. they had the mining logic already, just the `isclosed()` check was fucking it up

goliath's tetris-t attack and the melee retaliate ring both properly dig mineral turfs. the ranged retaliate chaser naturally stops upon encountering a turf so it does not spawn tentacles in mineral turfs and I think it's best left like this

Jacquerel commented on code/datums/ai/basic_mobs/basic_ai_behaviors/targetting.dm:
```diff
@@ -14,7 +14,7 @@
 
 	var/atom/current_target = controller.blackboard[target_key]
-	if (targetting_datum.can_attack(living_mob, current_target))
+	if (targetting_datum.can_attack(living_mob, current_target) && can_see(living_mob, current_target, vision_range))
```
Why is this in `targetting` and not in `can_attack`?
You could then also more easily make it configurably optional per mob.

Sealed101 committed:
power outage moment
moves it to can_attack (worms are annoying again)
targetting_datums now have a var for ignoring sight when pursuing a target, defaulted to FALSE
also bileworms use their own targetting_datum now (which only makes em ignore sight for targetting lol)
